### PR TITLE
Refactoring the User Management API

### PR DESCRIPTION
**Pull request by hiranya911**
Following changes were suggested at the API review:
* Introduce `UserInfo` public interface type
* Rename `User` to `UserRecord`
* Rename `User.Builder` to `UserRecord.CreateRequest`
* Rename `User.Updater` to `UserRecord.UpdateRequest`
* Drop `ProviderUserInfo` from the public API
* Implement `UserInfo` at both `UserRecord` and `ProviderUserInfo`

### CODE REVIEW

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 1]
```diff
@@ -39,51 +38,22 @@
     this.providerId = response.getProviderId();
   }
 
-  /**
-   * Returns the user's unique ID assigned by the identity provider.
-   *
-   * @return a user ID string.
-   */
   public String getUid() {
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Should these be annotated with @ Override?

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java [hunk 2]
```diff
@@ -39,51 +38,22 @@
...
     return uid;
   }
 
-  /**
-   * Returns the user's display name.
-   *
-   * @return a display name string or null.
-   */
-  @Nullable
   public String getDisplayName() {
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Not sure, but I think we still want @ Nullable annotation?

File: src/main/java/com/google/firebase/auth/UserRecord.java [hunk 3]
```diff
@@ -147,37 +163,13 @@ public UserMetadata getUserMetadata() {
   }
 
   /**
-   * Returns a new {@link User.Updater} instance, which can be used to update the attributes
+   * Returns a new {@link UpdateRequest}, which can be used to getProperties the attributes
```
Comment (mikelehen) [inline src/main/java/com/google/firebase/auth/UserRecord.java]: "used to getProperties" ?

File: src/main/java/com/google/firebase/auth/FirebaseAuth.java [hunk 4]
```diff
@@ -227,61 +229,62 @@ public User then(Task<GetTokenResult> task) throws Exception {
...
+   * {@link CreateRequest}.
    *
-   * @param builder A non-null {@link User.Builder} instance.
-   * @return A {@link Task} which will complete successfully with a {@link User} instance
+   * @param request A non-null {@link CreateRequest} instance.
+   * @return A {@link Task} which will complete successfully with a {@link UserRecord} instance
    *     corresponding to the newly created account. If an error occurs while creating the user
    *     account, the task fails with a FirebaseAuthException.
-   * @throws NullPointerException if the provided builder is null.
+   * @throws NullPointerException if the provided user is null.
```
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/FirebaseAuth.java]: Done

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java (hunk 1 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Done

File: src/main/java/com/google/firebase/auth/ProviderUserInfo.java (hunk 2 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/ProviderUserInfo.java]: Done

File: src/main/java/com/google/firebase/auth/UserRecord.java (hunk 3 above)
Comment (hiranya911) [inline src/main/java/com/google/firebase/auth/UserRecord.java]: silly auto refactor! Fixed :)

**Comment by hiranya911** [comment]
Made the suggested changes.

**Comment by Solido** [comment]
+1

**Pull request merged**